README.md
# container_baseimage Cookbook

This cookbook includes recipes to prepare a Ubuntu/Debian baseimage for Docker/Rkt/LXC usage.

## Requirements

### Platforms

- Ubuntu 18.04+

May work with or without modification on other Debian derivatives.

### Chef

- Chef 13.3+

### Cookbooks

- None

## Recipes

### default

Configures a container friendly baseimage that covers many corner cases including PID 1 zombie reaping, runit, cron and syslog daemons as well as log rotation.

## Maintainers

This cookbook is maintained by Morton Jonuschat <m.jonuschat@mojocode.de>

## License

**Copyright:** 2018, Morton Jonuschat

```
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# container_baseimage Cookbook CHANGELOG

This file is used to list changes made in each version of the container_baseimage cookbook.

## v2.0.1 (2018-12-31)


recipes/default.rb
#
# Cookbook:: container_baseimage
# Recipe:: default
#
# Copyright:: 2018, Morton Jonuschat <m.jonuschat@mojocode.de>
#
# Licensed under the Apache License, Version 2.0 (the 'License');
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an 'AS IS' BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe 'container_baseimage::prepare'
include_recipe 'container_baseimage::system_services'
include_recipe 'container_baseimage::utilities'


recipes/prepare.rb
#
# Cookbook:: container_baseimage
# Recipe:: prepare
#
# Copyright:: 2018, Morton Jonuschat <m.jonuschat@mojocode.de>
#
# Licensed under the Apache License, Version 2.0 (the 'License');
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an 'AS IS' BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Ensure known /etc/apt/sources.list contents
template '/etc/apt/sources.list' do
  source 'sources.list.erb'
  owner 'root'
  group 'root'
  mode '0644'
end

apt_update 'refresh' do
  action :update
end

## Fix some issues with APT packages.
link '/sbin/initctl' do
  to '/bin/true'
end

## Replace the 'ischroot' tool to make it always return true.
link '/sbin/initctl' do
  to '/bin/true'
end

## Install HTTPS support for APT, add-apt-repository and apt-utils
package %w[apt-transport-https apt-utils ca-certificates software-properties-common] do
  action :upgrade
end

## Ensure the en_US locale is available
package 'language-pack-en' do
  action :upgrade
end

locale 'en_US' do
  lang node['container_baseimage']['locale']['lang']
  lc_all node['container_baseimage']['locale']['lc_ctype']
  action :update
end

## Configure container environment variables
directory '/etc/container_environment' do
  action :create
  owner 'root'
  group 'root'
  mode '0700'
end

file '/etc/container_environment/LANG' do
  content node['container_baseimage']['locale']['lang']
  owner 'root'
  group 'root'
  mode '0644'
end

file '/etc/container_environment/LC_CTYPE' do
  content node['container_baseimage']['locale']['lc_ctype']
  owner 'root'
  group 'root'
  mode '0644'
end


recipes/utilities.rb
#
# Cookbook:: container_baseimage
# Recipe:: utilities
#
# Copyright:: 2018, Morton Jonuschat <m.jonuschat@mojocode.de>
#
# Licensed under the Apache License, Version 2.0 (the 'License');
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an 'AS IS' BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

## Often used tools.
package %w[curl less vim-tiny psmisc gpg-agent dirmngr] do
  action :upgrade
end

link '/usr/bin/vim' do
  to '/usr/bin/vim.tiny'
end

## This tool runs a command as another user and sets $HOME
cookbook_file '/sbin/setuser' do
  source 'setuser'
  owner 'root'
  group 'root'
  mode '0755'
  action :create
end

## This tool allows installation of apt packages with automatic cache cleanup.
cookbook_file '/sbin/install_clean' do
  source 'install_clean'
  owner 'root'
  group 'root'
  mode '0755'
  action :create
end


recipes/system_services.rb
#
# Cookbook:: container_baseimage
# Recipe:: system_services
#
# Copyright:: 2018, Morton Jonuschat <m.jonuschat@mojocode.de>
#
# Licensed under the Apache License, Version 2.0 (the 'License');
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an 'AS IS' BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

## Install init process
cookbook_file '/sbin/my_init' do
  source 'my_init'
  owner 'root'
  group 'root'
  mode '0755'
  action :create
end

%w[/etc/my_init.d /etc/my_init.pre_shutdown.d /etc/my_init.post_shutdown.d].each do |path|
  directory path do
    owner 'root'
    group 'root'
    mode '0755'
    action :create
  end
end

group 'docker_env' do
  gid 8377
end

%w[/etc/container_environment.sh /etc/container_environment.json].each do |path|
  file path do
    content ''
    owner 'root'
    group 'docker_env'
    mode '0640'
  end
end

link '/etc/profile.d/container_environment.sh' do
  to '/etc/container_environment.sh'
end

package %w[runit tzdata] do
  action :upgrade
end

## Install syslog-ng
if node['container_baseimage']['system_services']['syslog']
  package %w[syslog-ng-core logrotate] do
    action :upgrade
  end

  cookbook_file '/etc/logrotate.conf' do
    source 'logrotate/logrotate.conf'
    owner 'root'
    group 'root'
    mode '0640'
  end

  cookbook_file '/etc/logrotate.d/syslog-ng' do
    source 'logrotate/services/syslog-ng.conf'
    owner 'root'
    group 'root'
    mode '0640'
  end

  cookbook_file '/etc/my_init.d/10_syslog-ng.init' do
    source 'syslog-ng/syslog-ng.init'
    owner 'root'
    group 'root'
    mode '0755'
  end

  cookbook_file '/etc/my_init.post_shutdown.d/10_syslog-ng.shutdown' do
    source 'syslog-ng/syslog-ng.shutdown'
    owner 'root'
    group 'root'
    mode '0755'
  end

  cookbook_file '/etc/default/syslog-ng' do
    source 'syslog-ng/syslog-ng.default'
    owner 'root'
    group 'root'
    mode '0755'
  end

  cookbook_file '/etc/syslog-ng/syslog-ng.conf' do
    source 'syslog-ng/syslog-ng.conf'
    owner 'root'
    group 'root'
    mode '0640'
  end

  directory '/var/lib/syslog-ng' do
    owner 'root'
    group 'root'
    mode '0755'
  end

  file '/var/log/syslog' do
    content ''
    owner 'root'
    group 'root'
    mode '0640'
    action :create_if_missing
  end
end

## Install cron
if node['container_baseimage']['system_services']['cron']
  package 'cron' do
    action :upgrade
  end

  cookbook_file '/etc/pam.d/cron' do
    source 'cron/cron.pam'
    owner 'root'
    group 'root'
    mode '0644'
  end

  file '/etc/crontab' do
    owner 'root'
    group 'root'
    mode '0600'
  end

  %w[
    /etc/cron.daily/standard
    /etc/cron.daily/upstart
    /etc/cron.daily/dpkg
    /etc/cron.daily/password
    /etc/cron.weekly/fstrim
  ].each do |path|
    file path do
      action :delete
    end
  end

  runit_service 'cron' do
    log false
  end
end
